### PR TITLE
Make PETScDMDAMesh unique_ids more unique

### PR DESCRIPTION
**Pull request by roystgnr**
This fixes one of the problems in #15193 - Elem unique_id() values coming out of this generator will no longer overlap Node unique_id() values, and unique_ids should continue to be unique after any subsequent mesh modification too.

### CODE REVIEW

File: modules/external_petsc_solver/src/mesh/PETScDMDAMesh.C [hunk 1]
```diff
@@ -387,6 +388,10 @@ build_cube_Quad4(UnstructuredMesh & mesh, DM da)
   // Already partitioned!
   mesh.skip_partitioning(true);
 
+  // We've already set our own unique_id values; now make sure that
+  // future mesh modifications use subsequent values.
+  mesh.set_next_unique_id(Mx * My + (Mx - 1) * (My - 1));
```
Comment (lindsayad) [inline modules/external_petsc_solver/src/mesh/PETScDMDAMesh.C:393]: this looks fancy

File: modules/external_petsc_solver/src/mesh/PETScDMDAMesh.C [hunk 2]
```diff
@@ -186,7 +186,8 @@ add_element_Quad4(DM da,
   Elem * elem = new Quad4;
   elem->set_id(elem_id);
   elem->processor_id() = pid;
-  elem->set_unique_id() = elem_id;
+  // Make sure our unique_id doesn't overlap any nodes'
```
Comment (lindsayad) [inline modules/external_petsc_solver/src/mesh/PETScDMDAMesh.C:189]: `nodes'`, is that the the derivative of the nodes? :smile:

**Review by lindsayad — APPROVED** [review]

File: modules/external_petsc_solver/src/mesh/PETScDMDAMesh.C [hunk 3]
```diff
@@ -186,7 +186,8 @@ add_element_Quad4(DM da,
...
+  elem->set_unique_id() = elem_id + (nx + 1) * (ny + 1);
```
Comment (fdkong) [inline modules/external_petsc_solver/src/mesh/PETScDMDAMesh.C:190]: So unique ID will be different from ID?

Comment (roystgnr) [reply modules/external_petsc_solver/src/mesh/PETScDMDAMesh.C:190]: Yes.  That's pretty common, e.g. when we reuse id values during AMR/C but never unique_id values.  It's never safe to assume the contrary.

**Pull request merged**